kv/dist_sender_test.go

```diff
@@ -1068,3 +1068,119 @@ func TestSequenceUpdateOnMultiRangeQueryLoop(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+// TestMultiRangeSplitEndTransaction verifies that when a chunk of batch looks
+// like it's going to be dispatched to more than one range, it will be split
+// up if it it contains EndTransaction.
+func TestMultiRangeSplitEndTransaction(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	g, s := makeTestGossip(t)
+	defer s()
+
+	if err := g.SetNodeDescriptor(&roachpb.NodeDescriptor{NodeID: 1}); err != nil {
+		t.Fatal(err)
+	}
+	nd := &roachpb.NodeDescriptor{
+		NodeID:  roachpb.NodeID(1),
+		Address: util.MakeUnresolvedAddr(testAddress.Network(), testAddress.String()),
+	}
+	if err := g.AddInfoProto(gossip.MakeNodeIDKey(roachpb.NodeID(1)), nd, time.Hour); err != nil {
+		t.Fatal(err)
+
+	}
+
+	// Fill mockRangeDescriptorDB with two descriptors.
+	var descriptor1 = roachpb.RangeDescriptor{
+		RangeID:  1,
+		StartKey: roachpb.RKeyMin,
+		EndKey:   roachpb.RKey("b"),
+		Replicas: []roachpb.ReplicaDescriptor{
+			{
+				NodeID:  1,
+				StoreID: 1,
+			},
+		},
+	}
+	var descriptor2 = roachpb.RangeDescriptor{
+		RangeID:  2,
+		StartKey: roachpb.RKey("b"),
+		EndKey:   roachpb.RKeyMax,
+		Replicas: []roachpb.ReplicaDescriptor{
+			{
+				NodeID:  1,
+				StoreID: 1,
+			},
+		},
+	}
+	descDB := mockRangeDescriptorDB(func(key roachpb.RKey, _, _ bool) ([]roachpb.RangeDescriptor, error) {
+		desc := descriptor1
+		if !key.Less(roachpb.RKey("b")) {
+			desc = descriptor2
+		}
+		return []roachpb.RangeDescriptor{desc}, nil
+	})
+
+	for _, test := range []struct {
+		put1, put2, et roachpb.Key
+		// -1 is used as separator between requests for simplicity.
+		exp [][]roachpb.Method
+	}{
+		{
+			// Everything hits the first range, so we get a 1PC txn.
+			roachpb.Key("a1"), roachpb.Key("a2"), roachpb.Key("a3"),
+			[][]roachpb.Method{{roachpb.Put, roachpb.Put, roachpb.EndTransaction}},
+		},
+		{
+			// Only EndTransaction hits the second range.
+			roachpb.Key("a1"), roachpb.Key("a2"), roachpb.Key("b"),
+			[][]roachpb.Method{{roachpb.Put, roachpb.Put}, {roachpb.EndTransaction}},
+		},
+		{
+			// One write hits the second range, so EndTransaction has to be split off.
+			// In this case, going in the usual order without splitting off
+			// would actually be fine, but it doesn't seem worth optimizing at
+			// this point.
+			roachpb.Key("a1"), roachpb.Key("b1"), roachpb.Key("a1"),
+			[][]roachpb.Method{{roachpb.Put, roachpb.Noop}, {roachpb.Noop, roachpb.Put}, {roachpb.EndTransaction}},
+		},
+		{
+			// Both writes go to the second range, but not EndTransaction.
+			roachpb.Key("b1"), roachpb.Key("b2"), roachpb.Key("a1"),
+			[][]roachpb.Method{{roachpb.Put, roachpb.Put}, {roachpb.EndTransaction}},
+		},
+	} {
```

Comment:
This brace is awfully far from the `for` statement. I think we generally prefer to make a local variable for test cases rather than defining them inline in the `for` statement.